src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+    {
+        // Handle error, and notify session failure to the commissioner application.
+        ChipLogError(Controller, "Failed to process the certificate chain request");
+        // TODO: Map error status to correct error code
+        commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+    }
+}
+
+CHIP_ERROR DeviceCommissioner::ProcessCertificateChain(const ByteSpan & certificate)
+{
+    Transport::AdminPairingInfo * admin = mAdmins.FindAdminWithId(mAdminId);
+    VerifyOrReturnError(admin != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    switch (mCertificateChainBeingRequested)
+    {
+    case CertificateChainType::kDAC: {
+        admin->SetDeviceDAC(certificate);
+        break;
+    }
+    case CertificateChainType::kPAI: {
+        admin->SetDevicePAI(certificate);
+        break;
+    }
+    case CertificateChainType::kUnknown:
+    default: {
+        return CHIP_ERROR_INTERNAL;
+    }
+    }
+
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
+
+    Device * device = &mActiveDevices[mDeviceBeingPaired];
+
+    if (admin->AreDeviceCredentialsAvailable())
+    {
+        // Step c
+        // TODO: see if vid matches
+
+        // Step b/d
+        uint16_t paiLen;
+        const uint8_t * pai = admin->GetDevicePAI(paiLen);
+        uint16_t dacLen;
+        const uint8_t * dac = admin->GetDeviceDAC(dacLen);
+
+        ReturnErrorOnFailure(ValidateCertificateChain(paa_certificate, sizeof(paa_certificate), pai, paiLen, dac, dacLen));
+
+        // Extract Manufacturer Certificate Public Key
+        ReturnErrorOnFailure(ExtractPubkeyFromX509Cert(ByteSpan(dac, dacLen), mRemoteManufacturerPubkey));
+
+        ChipLogProgress(Controller, "Sending Attestation Request to the device.");
+        ReturnErrorOnFailure(SendAttestationRequestCommand(device));
+    }
+    else
+    {
+        CHIP_ERROR err = SendCertificateChainRequestCommand(device, CertificateChainType::kPAI);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(Controller, "Failed in sending Certificate Chain request command to the device: err %s", ErrorStr(err));
+            OnSessionEstablishmentError(err);
+            return err;
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceCommissioner::SendAttestationRequestCommand(Device * device)
+{
+    ChipLogDetail(Controller, "Sending Attestation request to %p device", device);
+    VerifyOrReturnError(device != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    chip::Controller::OperationalCredentialsCluster cluster;
+    cluster.Associate(device, 0);
+
+    Callback::Cancelable * successCallback = mAttestationResponseCallback.Cancel();
+    Callback::Cancelable * failureCallback = mOnAttestationFailureCallback.Cancel();
+
+    ReturnErrorOnFailure(Crypto::DRBG_get_bytes(mAttestationNonce, sizeof(mAttestationNonce)));
```

Comment:
The caller should be able to decide of the Nonce. This is necessary because the NOCResponse also has device attestation signature, and the goal of the AttestationNonce is that it should be possible to provide it from the same entity that validates signature. The methodology used here can work in the meantime, but is inflexible with regards to different commissioner policies. Suggest adding a TODO.

Reply:
Updated to mimic CSRNonce